File: CRM-Client/src/utils/licenseStatus.ts
export const LICENSE_STATUS_NONE = 'none'
export const LICENSE_STATUS_EXPIRED = 'expired'
export const LICENSE_STATUS_TRIAL = 'trial'
export const LICENSE_STATUS_OFFICIAL = 'official'

export type LicenseStatusValue =
  | typeof LICENSE_STATUS_NONE
  | typeof LICENSE_STATUS_EXPIRED
  | typeof LICENSE_STATUS_TRIAL
  | typeof LICENSE_STATUS_OFFICIAL

export const LICENSE_STATUS_LABELS: Record<LicenseStatusValue, string> = {
  none: '未授权',
  expired: '已过期',
  trial: '试用',
  official: '正式'
}

function toDateOnly(value: string | Date): Date | null {
  if (value instanceof Date) {
    if (Number.isNaN(value.getTime())) return null
    return new Date(value.getFullYear(), value.getMonth(), value.getDate())
  }
  const text = value.trim()
  if (text === '') return null
  const date = new Date(`${text.slice(0, 10)}T00:00:00`)
  if (Number.isNaN(date.getTime())) return null
  return date
}

export function classifyLicenseStatus(
  expiryDate: string | Date | null | undefined,
  licenseType: string | null | undefined,
  today: Date = new Date()
): LicenseStatusValue {
  if (expiryDate === null || expiryDate === undefined || expiryDate === '') {
    return LICENSE_STATUS_NONE
  }
  const expiry = toDateOnly(expiryDate)
  if (expiry === null) return LICENSE_STATUS_NONE
  const todayDate = new Date(today.getFullYear(), today.getMonth(), today.getDate())
  if (expiry < todayDate) return LICENSE_STATUS_EXPIRED
  if (licenseType === 'TRIAL') return LICENSE_STATUS_TRIAL
  return LICENSE_STATUS_OFFICIAL
}

export function licenseStatusLabel(
  expiryDate: string | Date | null | undefined,
  licenseType: string | null | undefined,
  today?: Date
): string {
  return LICENSE_STATUS_LABELS[classifyLicenseStatus(expiryDate, licenseType, today)]
}

export function licenseStatusClass(
  expiryDate: string | Date | null | undefined,
  licenseType: string | null | undefined,
  today?: Date
): string {
  return `license-badge--${classifyLicenseStatus(expiryDate, licenseType, today)}`
}

File: CRM-Server/app/core/list_query/license_status.py
from __future__ import annotations

from datetime import date

from sqlalchemy import case

LICENSE_STATUS_NONE = "none"
LICENSE_STATUS_EXPIRED = "expired"
LICENSE_STATUS_TRIAL = "trial"
LICENSE_STATUS_OFFICIAL = "official"

LICENSE_STATUS_VALUES = (
    LICENSE_STATUS_NONE,
    LICENSE_STATUS_EXPIRED,
    LICENSE_STATUS_TRIAL,
    LICENSE_STATUS_OFFICIAL,
)


def classify_license_status(
    expiry_date: date | None,
    license_type: str | None,
    today: date,
) -> str:
    if expiry_date is None:
        return LICENSE_STATUS_NONE
    if expiry_date < today:
        return LICENSE_STATUS_EXPIRED
    if license_type == "TRIAL":
        return LICENSE_STATUS_TRIAL
    return LICENSE_STATUS_OFFICIAL


def license_status_expression(expiry_column, type_column, today: date):
    return case(
        (expiry_column.is_(None), LICENSE_STATUS_NONE),
        (expiry_column < today, LICENSE_STATUS_EXPIRED),
        (type_column == "TRIAL", LICENSE_STATUS_TRIAL),
        else_=LICENSE_STATUS_OFFICIAL,
    )
